File: Walktrough_UnitTesting-ObjectiveC/Walktrough_UnitTesting-ObjectiveC/ViewController.m
//
//  ViewController.m
//  Walktrough_UnitTesting-ObjectiveC
//
//  Created by MCS Devices on 11/2/17.
//  Copyright © 2017 Mobile Consulting Solutions. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.dotNotation = NO;
    self.userIsInTheMiddleOfTyping = NO;
    self.firstEntered = NO;
    [self layoutComponents];
}

- (CalculatorBrain*) brain
{
    if(!_brain) _brain = [[CalculatorBrain alloc]init];
    return _brain;
}


- (void) layoutComponents {
    _inputLabel.layer.cornerRadius = 9;
    _inputLabel.layer.masksToBounds = true;
    _resultLabel.layer.cornerRadius = 9;
    _resultLabel.layer.masksToBounds = true;
    _oneButton.layer.cornerRadius = 7;
    _oneButton.layer.masksToBounds = true;
    _twoButton.layer.cornerRadius = 7;
    _twoButton.layer.masksToBounds = true;
    _threeButton.layer.cornerRadius = 7;
    _threeButton.layer.masksToBounds = true;
    _fourButton.layer.cornerRadius = 7;
    _fourButton.layer.masksToBounds = true;
    _fiveButton.layer.cornerRadius = 7;
    _fiveButton.layer.masksToBounds = true;
    _sixButton.layer.cornerRadius = 7;
    _sixButton.layer.masksToBounds = true;
    _sevenButton.layer.cornerRadius = 7;
    _sevenButton.layer.masksToBounds = true;
    _eightButton.layer.cornerRadius = 7;
    _eightButton.layer.masksToBounds = true;
    _nineButton.layer.cornerRadius = 7;
    _nineButton.layer.masksToBounds = true;
    _zeroButton.layer.cornerRadius = 7;
    _zeroButton.layer.masksToBounds = true;
    _dotButton.layer.cornerRadius = 7;
    _dotButton.layer.masksToBounds = true;
    _clearButton.layer.cornerRadius = 7;
    _clearButton.layer.masksToBounds = true;
    _plusButton.layer.cornerRadius = 7;
    _plusButton.layer.masksToBounds = true;
    _subtractButton.layer.cornerRadius = 7;
    _subtractButton.layer.masksToBounds = true;
    _divideButton.layer.cornerRadius = 7;
    _divideButton.layer.masksToBounds = true;
    _multiplyButton.layer.cornerRadius = 7;
    _multiplyButton.layer.masksToBounds = true;
    _enterButton.layer.cornerRadius = 7;
    _enterButton.layer.masksToBounds = true;
    
}

- (IBAction)digitPressed:(UIButton *)sender {
    
    NSString* digit = sender.currentTitle;
    
    if(!(self.dotNotation && [digit isEqualToString:@"."]))
    {
        if([digit isEqualToString:@"."])
            self.dotNotation=YES;
        
        if(self.userIsInTheMiddleOfTyping) //MUST have self
            self.inputLabel.text = [self.inputLabel.text stringByAppendingString:digit];
        else
        {
            self.inputLabel.text=digit;
            self.userIsInTheMiddleOfTyping=YES;
        }
    }
}

- (IBAction)operationPressed:(UIButton *)sender {
    
    
    if([sender.currentTitle isEqualToString:@"C"])
    {
        self.inputLabel.text = @"0";
        self.resultLabel.text=@"0";
        self.brain = nil;
        self.dotNotation = NO;
        self.userIsInTheMiddleOfTyping = NO;
        self.firstEntered = NO;
    }
    else
    {
    
        if(self.userIsInTheMiddleOfTyping) {
            [self enterPressed:nil];
        }
        self.inputLabel.text=[self.inputLabel.text stringByAppendingString:@" "];
        self.resultLabel.text=[self.resultLabel.text stringByAppendingString:sender.currentTitle];
        
        double result = [self.brain performOperation:sender.currentTitle];
        NSString * resultString = [NSString stringWithFormat:@"%g",result];
        self.inputLabel.text = resultString;
    }
}

- (IBAction)enterPressed:(UIButton *)sender {
    
    [self.brain pushOperand:[self.inputLabel.text doubleValue]];
    
    if(self.firstEntered)
    {
        self.resultLabel.text=[self.resultLabel.text stringByAppendingString:@" "];
        self.resultLabel.text=[self.resultLabel.text stringByAppendingString:self.inputLabel.text];
    }
    else
        self.resultLabel.text=self.inputLabel.text;
    
    self.userIsInTheMiddleOfTyping = NO;
    self.dotNotation=NO;
    self.firstEntered=YES;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Walktrough-UnitTesting-Swift/Walktrough-UnitTesting-Swift/ViewController.swift
//
//  ViewController.swift
//  Walktrough-UnitTesting-Swift
//
//  Created by MCS Devices on 11/2/17.
//  Copyright © 2017 Mobile Consulting Solutions. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var oneButton: UIButton!
    @IBOutlet weak var twoButton: UIButton!
    @IBOutlet weak var threeButton: UIButton!
    @IBOutlet weak var fourButton: UIButton!
    @IBOutlet weak var fiveButton: UIButton!
    @IBOutlet weak var sixButton: UIButton!
    @IBOutlet weak var sevenButton: UIButton!
    @IBOutlet weak var eightButton: UIButton!
    @IBOutlet weak var nineButton: UIButton!
    @IBOutlet weak var dotButton: UIButton!
    @IBOutlet weak var zeroButton: UIButton!
    @IBOutlet weak var clearButton: UIButton!
    @IBOutlet weak var plusButton: UIButton!
    @IBOutlet weak var substractButton: UIButton!
    @IBOutlet weak var divideButton: UIButton!
    @IBOutlet weak var multiplyButton: UIButton!
    @IBOutlet weak var enterButton: UIButton!
    @IBOutlet weak var resultLabel: UILabel!
    @IBOutlet weak var inputLabel: UILabel!
    
    
    var dotNotation:Bool?
    var userIsInTheMiddleOfTyping:Bool?
    var firstEntered: Bool?
    var brain: CalculatorBrain?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        layoutComponents()
        dotNotation = false;
        userIsInTheMiddleOfTyping = false;
        firstEntered = false;
        brain = CalculatorBrain()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    func layoutComponents() {
        inputLabel.layer.cornerRadius = 9;
        inputLabel.layer.masksToBounds = true;
        resultLabel.layer.cornerRadius = 9;
        resultLabel.layer.masksToBounds = true;
        oneButton.layer.cornerRadius = 7;
        oneButton.layer.masksToBounds = true;
        twoButton.layer.cornerRadius = 7;
        twoButton.layer.masksToBounds = true;
        threeButton.layer.cornerRadius = 7;
        threeButton.layer.masksToBounds = true;
        fourButton.layer.cornerRadius = 7;
        fourButton.layer.masksToBounds = true;
        fiveButton.layer.cornerRadius = 7;
        fiveButton.layer.masksToBounds = true;
        sixButton.layer.cornerRadius = 7;
        sixButton.layer.masksToBounds = true;
        sevenButton.layer.cornerRadius = 7;
        sevenButton.layer.masksToBounds = true;
        eightButton.layer.cornerRadius = 7;
        eightButton.layer.masksToBounds = true;
        nineButton.layer.cornerRadius = 7;
        nineButton.layer.masksToBounds = true;
        zeroButton.layer.cornerRadius = 7;
        zeroButton.layer.masksToBounds = true;
        dotButton.layer.cornerRadius = 7;
        dotButton.layer.masksToBounds = true;
        clearButton.layer.cornerRadius = 7;
        clearButton.layer.masksToBounds = true;
        plusButton.layer.cornerRadius = 7;
        plusButton.layer.masksToBounds = true;
        substractButton.layer.cornerRadius = 7;
        substractButton.layer.masksToBounds = true;
        divideButton.layer.cornerRadius = 7;
        divideButton.layer.masksToBounds = true;
        multiplyButton.layer.cornerRadius = 7;
        multiplyButton.layer.masksToBounds = true;
        enterButton.layer.cornerRadius = 7;
        enterButton.layer.masksToBounds = true;
    }

    @IBAction func digitPressed(sender: UIButton!){
    
        if let digit = sender.currentTitle {
    
            if(!(self.dotNotation! && digit == "."))
            {
                if(digit == "."){ self.dotNotation=true }
                
                if(self.userIsInTheMiddleOfTyping)!, let i = self.inputLabel.text {
                    self.inputLabel.text = "\(i)\(digit)"
                }
                else {
                    self.inputLabel.text = digit;
                    self.userIsInTheMiddleOfTyping=true;
                }
            }
        }
    }
    
    @IBAction func operationPressed(sender: UIButton!){
        if(sender.currentTitle == "C")
        {
            self.inputLabel.text = "0"
            self.resultLabel.text = "0"
            self.brain = CalculatorBrain()
            self.dotNotation = false;
            self.userIsInTheMiddleOfTyping = false
            self.firstEntered = false
        }
        else
        {
            if let s = sender.currentTitle, let i = self.inputLabel.text, let r = self.resultLabel.text {
                
                if(self.userIsInTheMiddleOfTyping)! {
                    self.enterPressed(sender: nil)
                }
                
                self.inputLabel.text = "\(i) "
                self.resultLabel.text = "\(r)\(i)\(s)"
                
                
                if let result = self.brain?.performOperation(operation: s){
                    let resultString = "\(result)"
                    self.inputLabel.text = resultString;
                }
            }
        }
    }
    
    @IBAction func enterPressed(sender: UIButton!){
        
        if let operand = Double(self.inputLabel.text!){
            self.brain?.pushOperand(operand: operand)
        
            if (self.firstEntered)!, let r = self.resultLabel.text, let i = self.inputLabel.text
            {
                self.resultLabel.text = "\(r) "
                self.resultLabel.text = "\(r)\(i)"
            }
            else if let i = self.inputLabel.text {
                self.resultLabel.text = i
                self.inputLabel.text = "0"
            }
            self.userIsInTheMiddleOfTyping = false
            self.dotNotation = false
            self.firstEntered = true
        }
    }

}

